### PR TITLE
Support running different allocation analyses

### PR DESCRIPTION
**Pull request by victor-eds**
Each of the two commits in this PR should be reviewed separately.

08786aca26d5a19f85f14d3148afb8a661ebbc96: To be upstreamed. Extends memory allocation analysis infrastructure so the analysis to use can be configured.

d3fe27bba0254948a877b1ea2e60cfd91224ea58: Downstream. Use an Intel-specific analysis. This is just a copy of the original for now, so, please, *do not review the analysis implementation*.

96a2e78f1df431a52e989affb95b90c10980c358 and cbfdaf9a000d11e157d7fad1d1cea8e1e4d2b75c should be part of the first commit. Will squash there when upstreaming.

### CODE REVIEW

File: include/triton/Analysis/Allocation.h [hunk 1]
```diff
@@ -250,17 +257,19 @@ class ModuleAllocation : public CallGraph<Allocation> {
 public:
   using FuncOffsetMapT = DenseMap<FunctionOpInterface, Value>;
 
-  explicit ModuleAllocation(ModuleOp moduleOp)
-      : CallGraph<Allocation>(moduleOp) {
-    walk<WalkOrder::PreOrder, WalkOrder::PostOrder>(
+  template <typename AllocationAnalysis = triton::AllocationAnalysis>
+  static ModuleAllocation get(ModuleOp moduleOp) {
+    ModuleAllocation res(moduleOp);
+    res.walk<WalkOrder::PreOrder, WalkOrder::PostOrder>(
```
Comment (victor-eds) [inline include/triton/Analysis/Allocation.h:263]: Cannot have templated constructors like this, so this is a static member function now.

File: third_party/intel/lib/Analysis/Allocation.cpp
Comment (victor-eds) [inline third_party/intel/lib/Analysis/Allocation.cpp:1]: Do not review analysis implementation. This is a copy of [Allocation.cpp](https://github.com/intel/intel-xpu-backend-for-triton/blob/0df7d80e8a4e46a5cdb2f729a4d3c0d9d67b0a3c/lib/Analysis/Allocation.cpp).

File: include/triton/Analysis/Allocation.h [hunk 2]
```diff
@@ -238,6 +241,7 @@ class Allocation {
   size_t sharedMemorySize = 0;
 
   friend class triton::AllocationAnalysis;
+  friend class triton::intel::AllocationAnalysis;
```
Comment (chengjunlu) [inline include/triton/Analysis/Allocation.h:244]: The change of `intel` namespace in a public header enlarge the divergence of the code upstream and downstream.

Can we improve this?

**Comment by chengjunlu** [comment]
Does this PR solves the issue about the allocation https://github.com/intel/intel-xpu-backend-for-triton/issues/2340

**Comment by victor-eds** [comment]
> Does this PR solves the issue about the allocation https://github.com/intel/intel-xpu-backend-for-triton/issues/2340

This is a first PR refactoring the code so we can modify freely and fix issues like that one and the ones I'm finding with my current FA work

File: include/triton/Analysis/Allocation.h (hunk 2 above)
Comment (victor-eds) [reply include/triton/Analysis/Allocation.h:244]: Good point. I'll further rector with that in mind.

Comment (etiotto) [reply include/triton/Analysis/Allocation.h:244]: That means no concrete allocation analysis can be a friend of this class probably and there will need to be public member function to access all the private member data fields of this class (which is not a bad thing IMO).

File: third_party/intel/lib/Analysis/Allocation.cpp [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Support/LLVM.h"
+#include "triton/Analysis/Alias.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include "triton/Dialect/Triton/IR/Utility.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "llvm/ADT/SmallVector.h"
+
+using ::mlir::triton::gpu::AMDMfmaEncodingAttr;
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/Allocation.cpp:18]: Remove this one

File: third_party/intel/lib/Analysis/Allocation.cpp [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+using ::mlir::triton::gpu::BlockedEncodingAttr;
+using ::mlir::triton::gpu::DotOperandEncodingAttr;
+using ::mlir::triton::gpu::getContigPerThread;
+using ::mlir::triton::gpu::getOrder;
+using ::mlir::triton::gpu::getShapePerCTA;
+using ::mlir::triton::gpu::getShapePerCTATile;
+using ::mlir::triton::gpu::getSizePerThread;
+using ::mlir::triton::gpu::getUniqueContigPerThread;
+using ::mlir::triton::gpu::NvidiaMmaEncodingAttr;
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/Allocation.cpp:27]: Remove

File: third_party/intel/lib/Analysis/Allocation.cpp [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+#include "triton/Analysis/Alias.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include "triton/Dialect/Triton/IR/Utility.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "llvm/ADT/SmallVector.h"
+
+using ::mlir::triton::gpu::AMDMfmaEncodingAttr;
+using ::mlir::triton::gpu::BlockedEncodingAttr;
+using ::mlir::triton::gpu::DotOperandEncodingAttr;
+using ::mlir::triton::gpu::getContigPerThread;
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/Allocation.cpp:21]: Unused?

File: third_party/intel/lib/Analysis/Allocation.cpp [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+using ::mlir::triton::gpu::getOrder;
+using ::mlir::triton::gpu::getShapePerCTA;
+using ::mlir::triton::gpu::getShapePerCTATile;
+using ::mlir::triton::gpu::getSizePerThread;
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/Allocation.cpp:25]: Unused?

File: third_party/intel/lib/Analysis/Allocation.cpp [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+using ::mlir::triton::gpu::getUniqueContigPerThread;
+using ::mlir::triton::gpu::NvidiaMmaEncodingAttr;
+using ::mlir::triton::gpu::SharedEncodingAttr;
+using ::mlir::triton::gpu::SliceEncodingAttr;
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/Allocation.cpp:29]: Unused?

File: third_party/intel/lib/Analysis/Allocation.cpp [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+// Shared Memory Allocation Analysis
+//===----------------------------------------------------------------------===//
+namespace triton::intel {
+
+// Bitwidth of pointers
+constexpr int kPtrBitWidth = 64;
+
+static std::pair<SmallVector<unsigned>, SmallVector<unsigned>>
+getCvtOrder(Attribute srcLayout, Attribute dstLayout) {
+  auto srcMmaLayout = mlir::dyn_cast<NvidiaMmaEncodingAttr>(srcLayout);
```
Comment (etiotto) [inline third_party/intel/lib/Analysis/Allocation.cpp:43]: This should be a DPAS layout right ? OR are you just copying the original file first and do further modification later ?

**Comment by victor-eds** [comment]
`Allocation.cpp` is just a copy for now. No need to review. Rest is ready to upstream.

**Review by chengjunlu — APPROVED** [review]

**Review by etiotto — APPROVED** [review]

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**